agent/websockets/connection.go

```diff
@@ -252,3 +264,60 @@ func (conn *Connection) ReadServerMessages() ([]interface{}, error) {
 		return nil, nil
 	}
 }
+
+// InjectedHeader represents a single header which is injected into a websocket message.
+type InjectedHeader struct {
+	// Header is the header to inject from.
+	Header string `json:"header,omitempty"`
+	// Path is a period-delimited string where each element refers to the child
+	// element of the previous element. For example, for the injection path `a.b.c`, it will
+	// inject the header value into the nested JSON objects for key c under key b under key a
+	// under the parent: `{"a": {"b": {"c": {"<key>": "<injected_header_value>"}}}}`
+	Path []string `json:"path,omitempty"`
+	// Key is the key for which to inject the value as.
+	Key string `json:"key,omitempty"`
+}
+
+// injectWebsocketMessage injects a shim header value into a single websocket message in-place.
+// Returns a pointer to a new copy of the struct on success.
+func injectWebsocketMessage(msg *message, injectionPath []string, injectionValues map[string]string) (*message, error) {
+	if msg == nil || injectionValues == nil {
+		return nil, errors.New("unexpected nil message")
+	}
+	if injectionValues == nil {
+		return nil, errors.New("unexpected nil values map")
+	}
+	if len(injectionValues) == 0 {
+		return msg, nil
+	}
+	// Deserialize the websocket message into a JSON object.
+	var origJSONComponent map[string]interface{}
+	err := json.Unmarshal(msg.Data, &origJSONComponent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal as json message: %v", err)
+	}
+	var currJSONComponent map[string]interface{}
+	var ok bool
```

Comment:
nit: no need to declare

Reply:
It's necessary because if we perform `:=` assignment in line 303, it will localize the scope of currJSONComponent to the for loop which would be undesired behavior.